PlasmaMVP/throughputOptimization.md

```diff
@@ -0,0 +1,52 @@
+### Increase Throughput of Plasma
+
+Rather than submitting a block to the Ethereum rootchain and waiting for presumed finality before submitting the next block; we propose changes to the submitBlock function that allow blocks to be added before the previous blocks get finalized. We effectively allow the blocks to be added "asynchronously".
+
+The submitBlock function changes to:
+``submitBlock(blockNum, root, prevRoot, sig)``
+
+blockNum: The blocknumber that is being submitted
+
+root: Merkle root of transaction tree in block
+
+prevRoot: Merkle root of transaction tree in previous block (blockNum - 1)
+
+sig: Signature by the authority over the hash of blockNum, root, prevRoot.
+
+The contract keeps track of the largest blocknumber seen, call this ``latestBlock`` as well as a childchain mapping from ``(uint blocknumber => Block struct)``
+
+For a given call of submitBlock:
+
+#### Case 1: Submitting a new Block
+If the blockNum is ``1 + latestBlock``, we simply check that ``ChildChain[latestBlock].root == prevRoot``
+
+#### Case 2: Filling in reorged blocks in the middle
+If a block (with blocknumber k < latestBlock) in the middle gets reorged out of existence, one can fill in the gap by creating ``submitBlock(k, kth_block_root, k-1th_block_root, sig)``
+
+Here we check that ``ChildChain[k + 1].prevRoot == root``, this prevents a malicious validator from submitting a block at height k that they did not originally submit in the case where kth block gets reorged.
+If multiple blocks are reorged in a row, we simply resubmit the missing block with highest missing height and move towards the lowest missing height.
+
+#### Case 3: n latest blocks get reorged out of existence
+Lastly, there is the case where the last n blocks get reorged out of existence. This is analogous to have one very big block getting reorged out of the existence, and the blocks must be added back in order from lowest height to highest height since latestBlock has been changed to ``latestBlock - n``
+
+In all cases, we check that ``sig`` is the authority's signature over ``keccak(blockNum, root, prevRoot)``. The reason we include this signature is so that anyone can submit blocks to the rootchain so long as they have the authority's signature. This prevents validators from refusing to fill in gaps in the case of a reorg.
```

Comment:
Might also be worth nothing that if a validator chooses not to broadcast the header and sig, it would be grounds to exit the child chain